In [17]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import operator
import math
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
%matplotlib inline

In [18]:
#df = pd.read_csv("/home/hp/Desktop/College/Sem5/DA/Project/zomato-bangalore-restaurants/zomato.csv")
mylist = []

for chunk in  pd.read_csv("C:/Users/sparsha/Documents/SEM 5/DA Project/zomato/newzomato.csv", sep=',', chunksize=20000):
    mylist.append(chunk)

big_data = pd.concat(mylist, axis= 0)
df = big_data

del mylist

Linear regression model with NA values filled with median

In [19]:
transformed_df = df.copy()
transformed_df['rate'].fillna((transformed_df['rate'].median()), inplace=True)
transformed_df['approx_cost(for two people)'].fillna((transformed_df['approx_cost(for two people)'].median()), inplace=True)

In [20]:
transformed_df['new_Cuisine1'], _ = pd.factorize(transformed_df['Cuisine1'], sort=True)
transformed_df['new_Cuisine2'], _ = pd.factorize(transformed_df['Cuisine2'], sort=True)
transformed_df['new_rest1'], _ = pd.factorize(transformed_df['rest1'], sort=True)
transformed_df['new_listed_in'], _ = pd.factorize(transformed_df['listed_in(type)'], sort=True)
transformed_df['new_location'], _ = pd.factorize(transformed_df['location'], sort=True)
transformed_df['new_online_order'], _ = pd.factorize(transformed_df['online_order'], sort=True)
transformed_df['new_book_table'], _ = pd.factorize(transformed_df['book_table'], sort=True)

to_drop = ['Cuisine1',
           'Cuisine2',
           'rest1',
           'listed_in(type)',
           'location',
           'online_order',
           'book_table',
           'name']
transformed_df.drop(to_drop, inplace=True, axis=1)
transformed_df.dropna(how='any', inplace=True)

In [21]:
train, test = train_test_split(transformed_df, test_size=0.2)

In [22]:
train_x = train.loc[:, train.columns != 'rate']
train_y = train.loc[:, train.columns == 'rate']
test_x = test.loc[:, test.columns != 'rate']
test_y = test.loc[:, test.columns == 'rate']

In [23]:
regressor = LinearRegression()  
regressor.fit(train_x, train_y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [24]:
y_pred = regressor.predict(test_x)

In [31]:
y_pred = np.around(y_pred,decimals=1)
#test_y = list(test_y['rate'])
acc = 0
for i in range(len(y_pred)):
    if(np.around(abs(y_pred[i] - test_y[i]),1) <= 0.1):
        acc = acc + 1
print(acc)        
print(acc/len(y_pred))

4218
0.4137322216772928


Linear regression model with NA values filled by linear interpolation

In [32]:
transformed_df2 = df.copy()
transformed_df2 = df.interpolate()

In [33]:
transformed_df2['new_Cuisine1'], _ = pd.factorize(transformed_df2['Cuisine1'], sort=True)
transformed_df2['new_Cuisine2'], _ = pd.factorize(transformed_df2['Cuisine2'], sort=True)
transformed_df2['new_rest1'], _ = pd.factorize(transformed_df2['rest1'], sort=True)
transformed_df2['new_listed_in'], _ = pd.factorize(transformed_df2['listed_in(type)'], sort=True)
transformed_df2['new_location'], _ = pd.factorize(transformed_df2['location'], sort=True)
transformed_df2['new_online_order'], _ = pd.factorize(transformed_df2['online_order'], sort=True)
transformed_df2['new_book_table'], _ = pd.factorize(transformed_df2['book_table'], sort=True)

to_drop = ['Cuisine1',
           'Cuisine2',
           'rest1',
           'listed_in(type)',
           'location',
           'online_order',
           'book_table',
           'name']
transformed_df2.drop(to_drop, inplace=True, axis=1)
transformed_df2.dropna(how='any', inplace=True)


In [34]:
train, test = train_test_split(transformed_df2, test_size=0.2)

In [35]:
train_x = train.loc[:, train.columns != 'rate']
train_y = train.loc[:, train.columns == 'rate']
test_x = test.loc[:, test.columns != 'rate']
test_y = test.loc[:, test.columns == 'rate']

In [36]:
regressor = LinearRegression()  
regressor.fit(train_x, train_y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [37]:
y_pred = regressor.predict(test_x)

In [39]:
y_pred = np.around(y_pred,decimals=1)
test_y = np.around(list(test_y['rate']),decimals=1)
acc = 0
for i in range(len(y_pred)):
    if(np.around(abs(y_pred[i] - test_y[i]),1) <= 0.1):
        acc = acc + 1
print(acc/len(y_pred))


0.33928396272682687


Linear regression model with NA values filled by quadratic interpolation

In [40]:
transformed_df3 = df.copy()
transformed_df3 = df.interpolate(method='polynomial', order=2)

In [41]:
transformed_df3['new_Cuisine1'], _ = pd.factorize(transformed_df3['Cuisine1'], sort=True)
transformed_df3['new_Cuisine2'], _ = pd.factorize(transformed_df3['Cuisine2'], sort=True)
transformed_df3['new_rest1'], _ = pd.factorize(transformed_df3['rest1'], sort=True)
transformed_df3['new_listed_in'], _ = pd.factorize(transformed_df3['listed_in(type)'], sort=True)
transformed_df3['new_location'], _ = pd.factorize(transformed_df3['location'], sort=True)
transformed_df3['new_online_order'], _ = pd.factorize(transformed_df3['online_order'], sort=True)
transformed_df3['new_book_table'], _ = pd.factorize(transformed_df3['book_table'], sort=True)

to_drop = ['Cuisine1',
           'Cuisine2',
           'rest1',
           'listed_in(type)',
           'location',
           'online_order',
           'book_table',
           'name']
transformed_df3.drop(to_drop, inplace=True, axis=1)
transformed_df3.dropna(how='any', inplace=True)


In [42]:
train, test = train_test_split(transformed_df3, test_size=0.2)


In [43]:
train_x = train.loc[:, train.columns != 'rate']
train_y = train.loc[:, train.columns == 'rate']
test_x = test.loc[:, test.columns != 'rate']
test_y = test.loc[:, test.columns == 'rate']


In [44]:
regressor = LinearRegression()  
regressor.fit(train_x, train_y)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [45]:
y_pred = regressor.predict(test_x)

In [46]:
y_pred = np.around(y_pred,decimals=1)
test_y = np.around(list(test_y['rate']),decimals=1)
acc = 0
for i in range(len(y_pred)):
    if(np.around(abs(y_pred[i] - test_y[i]),1)<= 0.1):
        acc = acc + 1
print(acc/len(y_pred))


0.30828837665522313


Accuracy of Linear regression model with NA values filled with median : 41.37%
Accuracy of Linear regression model with NA values filled by linear interpolation  : 33.92%
Accuracy of Linear regression model with NA values filled by linear interpolation  : 30.82%


Linear regression model ffilled with median has higher accuracy of around 41% compared to the other